### ISSUE TITLE
Improve error message: ABIEncoderV2 public functions with struct containing the tree of mapping (nested mapping)

### ISSUE THREAD

**Issue by strrel** [ethereum/solidity#4260]
ABIEncoderV2 public functions with struct containing the tree of mapping with struct. For example, this code returns the error:

> browser/test.sol:19:26: TypeError: Internal or recursive type is not allowed for public or external functions.
function addQuestion(MyStructName2 _x) returns (bool result) 
            ^--------------^

```
pragma solidity ^0.4.22;
pragma experimental ABIEncoderV2;

contract Test {
    struct MyStructName1 {
        address addr;
        uint256 count;
        mapping(uint => MyStructName2) x;
    }
    struct MyStructName2 {
        address addr;
        uint256 count;
        mapping(uint => MyStructName3) y;
    }
    struct MyStructName3 {
        address addr;
        uint256 count;
    }
    function addQuestion(MyStructName2 _x) returns (bool result) {
    }   
}
```
But here is no recursion. This code, for example doesn't return error:

```
pragma solidity ^0.4.22;
pragma experimental ABIEncoderV2;
contract Test {
    struct MyStructName1 {
        address addr;
        uint256 count;
        mapping(uint => MyStructName2) x;
    }
    struct MyStructName2 {
        address addr;
        uint256 count;
        mapping(uint => MyStructName3) y;
    }
    struct MyStructName3 {
        address addr;
        uint256 count;
    }
    function addQuestion(MyStructName3 _z) returns (bool result) {
    }
}
```

I believe recursive mappings looks like:
```
pragma solidity ^0.4.22;
pragma experimental ABIEncoderV2;
contract Test {
    struct MyStructName1 {
        address addr;
        uint256 count;
        mapping(uint => MyStructName1) x;
    }
    function addQuestion(MyStructName1 _x) returns (bool result) {
    ....
    }
}
```
So I think here are two options:
1) to accept using nested mapping (rather than recursive).
2) to throw the correct error, for example: "Nested types is not allowed for public or external functions."
But I think that there is no reasons to prohibit nested mappings.

**Comment by chriseth** [comment]
The issues is not about the nesting. The issue is that mappings can only live in storage.

We should improve the message to state this, though.

**Comment by chriseth** [comment]
This could use the template @bit-shift built for returning error messages in deep calls..

**Comment by Marenz** [comment]
@erak I unassigned you for now assuming you are not actively working on this. Feel free to re-assign if I was wrong :)

**Comment by chriseth** [comment]
Should be solved with #7739

**Comment by chriseth** [comment]
Solved by #9146
